src/main/java/net/snowflake/client/jdbc/cloud/storage/GcmEncryptionProvider.java

```diff
@@ -0,0 +1,235 @@
+ */
+package net.snowflake.client.jdbc.cloud.storage;
+
+import static java.nio.file.StandardOpenOption.CREATE;
+import static java.nio.file.StandardOpenOption.READ;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Base64;
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import net.snowflake.client.jdbc.MatDesc;
+import net.snowflake.common.core.RemoteStoreFileEncryptionMaterial;
+
+class GcmEncryptionProvider {
+  private static final int TAG_LENGTH = 128;
+  private static final String AES = "AES";
+  private static final String FILE_CIPHER = "AES/GCM/NoPadding";
+  private static final String KEY_CIPHER = "AES/GCM/NoPadding";
+  private static final int BUFFER_SIZE = 8 * 1024 * 1024; // 2 MB
+  private static final int blockSize;
+  private static final SecureRandom random;
+  private static final Base64.Decoder base64Decoder = Base64.getDecoder();
+
+  static {
+    try {
+      Cipher fileCipher = Cipher.getInstance(FILE_CIPHER);
+      blockSize = fileCipher.getBlockSize();
+
+      random = SecureRandom.getInstance("SHA1PRNG");
+    } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+      throw new ExceptionInInitializerError(e);
+    }
+  }
+
+  static InputStream encrypt(
+      StorageObjectMetadata meta,
+      long originalContentLength,
+      InputStream src,
+      RemoteStoreFileEncryptionMaterial encMat,
+      SnowflakeStorageClient client,
+      byte[] dataAad,
+      byte[] keyAad)
+      throws InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException,
+          BadPaddingException, NoSuchPaddingException, NoSuchAlgorithmException {
+
+    byte[] kek = base64Decoder.decode(encMat.getQueryStageMasterKey());
+    int keySize = kek.length;
+    byte[] keyBytes = new byte[keySize];
+    byte[] dataIvBytes = new byte[blockSize];
+    byte[] keyIvBytes = new byte[blockSize];
+    initRandomIvsAndFileKey(dataIvBytes, keyIvBytes, keyBytes);
+    byte[] encryptedKey = encryptKey(kek, keyBytes, keyIvBytes, keyAad);
+    CipherInputStream cis = encryptContent(src, keyBytes, dataIvBytes, dataAad);
+    addEncryptionMetadataToStorageClient(
+        meta,
+        originalContentLength,
+        encMat,
+        client,
+        keySize,
+        encryptedKey,
+        dataIvBytes,
+        keyIvBytes,
+        keyAad,
```

Comment:
as above?

Reply:
You mean AAD here? Yes, AAD is something related to GCM - one of the parameters of the algorithm.